frontend/server/src/main/java/org/pytorch/serve/util/GPUManager.java

```diff
@@ -0,0 +1,165 @@
+import java.io.InputStream;
+import java.io.InputStreamReader;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class GPUManager {
+
+    private static final Logger logger = LoggerFactory.getLogger(GPUManager.class);
+    private static final int nFailureHistory = 100;
+
+    private static GPUManager instance;
+
+    private final int nGPUs;
+    private final int minFreeMemory;
+    private final float maxShareFailures;
+
+    private AtomicInteger[] freeMemory;
+    private ConcurrentHashMap<String, Integer> workerIds;
+    private LinkedBlockingDeque<Integer> gpuFailureHistory;
+
+    private GPUManager(int nGPUs, int minFreeMemory, float maxShareFailures) {
+        this.nGPUs = nGPUs;
+        this.minFreeMemory = minFreeMemory;
+        this.maxShareFailures = maxShareFailures;
+
+        this.gpuFailureHistory = new LinkedBlockingDeque<> (nFailureHistory);
+        this.workerIds = new ConcurrentHashMap<> ();
+
+        if (nGPUs > 0) {
+            this.freeMemory = new AtomicInteger[this.nGPUs];
+            for (int i = 0; i < this.nGPUs; i++) {
+                this.freeMemory[i] = new AtomicInteger(-1);
+            }
+        }
+    }
+
+    // code largely copied from WorkerThread::getGpuUsage
+    private int queryNvidiaSmiFreeMemory(int gpuId) {
+        Process process;
+        try {
+            process =
+                    Runtime.getRuntime()
+                            .exec(
+                                    "nvidia-smi -i "
+                                            + gpuId
+                                            + " --query-gpu=memory.free --format=csv,noheader,nounits");
+            process.waitFor();
+            int exitCode = process.exitValue();
+            if (exitCode != 0) {
+                InputStream error = process.getErrorStream();
+                for (int i = 0; i < error.available(); i++) {
+                    logger.error("" + error.read());
+                }
+                return -1;
+            }
+            InputStream stdout = process.getInputStream();
+            BufferedReader reader =
+                    new BufferedReader(new InputStreamReader(stdout, StandardCharsets.UTF_8));
+            String line = reader.readLine();
+            if (line == null) {
+                return -1;
+            } else {
+                return Integer.parseInt(line);
+            }
+        } catch (Exception e) {
+            logger.error("An exception occurred when querying for free gpu memory", e);
+        }
+
+        return -1;
+    }
+
+    public static void init(ConfigManager configManager) {
+        int nGPUs = configManager.getNumberOfGpu();
+        int minFreeMemory = configManager.getMinFreeGpuMemory();
+        float maxShareFailures = configManager.getMaxShareGpuFailures();
+        instance = new GPUManager(nGPUs, minFreeMemory, maxShareFailures);
+    }
+
+    public static GPUManager getInstance() {
+        return instance;
+    }
```

Comment:
Can `init` also be called by multiple threads? Then maybe we should also make these static method synchronized - otherwise the manager might get constructed twice.